core/dbt/parser/schemas.py

```diff
@@ -1265,4 +1279,101 @@ def parse_patch(self, block: TargetBlock[UnparsedMacroUpdate], refs: ParserRef)
         macro.created_at = time.time()
         macro.meta = patch.meta
         macro.docs = patch.docs
-        macro.arguments = patch.arguments
+
+        if getattr(get_flags(), "validate_macro_args", False):
+            self._check_patch_arguments(macro, patch)
+            macro.arguments = patch.arguments if patch.arguments else macro.arguments
+        else:
+            macro.arguments = patch.arguments
+
+    def _check_patch_arguments(self, macro: Macro, patch: ParsedMacroPatch) -> None:
+        if not patch.arguments:
+            return
+
+        for macro_arg, patch_arg in zip(macro.arguments, patch.arguments):
+            if patch_arg.name != macro_arg.name:
+                msg = f"Argument {patch_arg.name} in yaml for macro {macro.name} does not match the jinja definition."
+                fire_event(Note(msg=msg), EventLevel.WARN)
```

Comment:
Are we intentionally doing `fire_event` here or should we be doing `warn_or_error`?